syntax = "proto3";

package wool._protobuf.mempool;

service MemoryPool {
    rpc session (SessionRequest) returns (stream SessionResponse);
    rpc acquire (AcquireRequest) returns (AcquireResponse);
    rpc map (AcquireRequest) returns (AcquireResponse);
    rpc put (PutRequest) returns (PutResponse);
    rpc post (PostRequest) returns (PostResponse);
    rpc get (GetRequest) returns (GetResponse);
    rpc release (ReleaseRequest) returns (ReleaseResponse);
}

message SessionRequest {
}

message SessionResponse {
    oneof response {
        Session session = 1;
        Event event = 2;
    }
}

message AcquireRequest {
    Reference reference = 1;
    Session session = 2;
}

message AcquireResponse {
}

message PutRequest {
    Session session = 1;
    bool mutable = 2;
    bytes dump = 3;
}

message PutResponse {
    Reference reference = 1;
}

message PostRequest {
    Session session = 1;
    Reference reference = 2;
    bytes dump = 3;
}

message PostResponse {
    bool updated = 1;
}

message GetRequest {
    Reference reference = 1;
    Session session = 2;
}

message GetResponse {
    bytes dump = 1;
}

message ReleaseRequest {
    Reference reference = 1;
    Session session = 2;
}

message ReleaseResponse {
}

message Reference {
    string id = 1;
}

message Session {
    string id = 1;
}

message Event {
    Reference reference = 1;
    string event_type = 2;
}
